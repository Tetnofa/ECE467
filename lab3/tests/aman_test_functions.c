// Test function declarations, calls and recursion
int factorial(int n) {
    if (n <= 1) {
        return 1;
    }
    return factorial(n - 1);
}

bool isEven(int num) {
    if (num == 0) {
        return true;
    }
    if (num == 1) {
        return false;
    }
    return isEven(num - 2);
}

void testFunctions() {
    int result;
    bool evenTest;
    
    result = factorial(5);
    writeInt(result);
    newLine();
    
    evenTest = isEven(4);
    writeBool(evenTest);
    newLine();
} 