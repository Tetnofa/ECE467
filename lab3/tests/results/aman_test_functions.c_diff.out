0a1,112
> Program [useIO=0](2,0)
> 	Function[isProto=0]	Int(2,0)
> 		Identifier[name:factorial](2,4)
> 		Parameter		Int(2,14)
> 			Identifier[name:n](2,18)
> 		Scope(2,21)
> 			IfStmt[hasElse=0](3,4)
> 				Bool Expression(3,8)
> 					Binary Expression[opcode:<=](3,8)
> 						Int Expression(3,8)
> 							Reference(3,8)
> 								Identifier[name:n](3,8)
> 						Int Expression(3,13)
> 							IntConstant[val=1](3,13)
> 				Scope(3,16)
> 					ReturnStmt(4,8)
> 						Int Expression(4,15)
> 							IntConstant[val=1](4,15)
> 			ReturnStmt(6,4)
> 				CallExpr(6,11)
> 					Identifier[name:factorial](6,11)
> 					Argument(6,21)
> 						Int Expression(6,21)
> 							Binary Expression[opcode:-](6,21)
> 								Int Expression(6,21)
> 									Reference(6,21)
> 										Identifier[name:n](6,21)
> 								Int Expression(6,25)
> 									IntConstant[val=1](6,25)
> 	Function[isProto=0]	Bool(9,0)
> 		Identifier[name:isEven](9,5)
> 		Parameter		Int(9,12)
> 			Identifier[name:num](9,16)
> 		Scope(9,21)
> 			IfStmt[hasElse=0](10,4)
> 				Bool Expression(10,8)
> 					Binary Expression[opcode:==](10,8)
> 						Int Expression(10,8)
> 							Reference(10,8)
> 								Identifier[name:num](10,8)
> 						Int Expression(10,15)
> 							IntConstant[val=0](10,15)
> 				Scope(10,18)
> 					ReturnStmt(11,8)
> 						Int Expression(11,15)
> 							BoolConstant[val=1](11,15)
> 			IfStmt[hasElse=0](13,4)
> 				Bool Expression(13,8)
> 					Binary Expression[opcode:==](13,8)
> 						Int Expression(13,8)
> 							Reference(13,8)
> 								Identifier[name:num](13,8)
> 						Int Expression(13,15)
> 							IntConstant[val=1](13,15)
> 				Scope(13,18)
> 					ReturnStmt(14,8)
> 						Int Expression(14,15)
> 							BoolConstant[val=0](14,15)
> 			ReturnStmt(16,4)
> 				CallExpr(16,11)
> 					Identifier[name:isEven](16,11)
> 					Argument(16,18)
> 						Int Expression(16,18)
> 							Binary Expression[opcode:-](16,18)
> 								Int Expression(16,18)
> 									Reference(16,18)
> 										Identifier[name:num](16,18)
> 								Int Expression(16,24)
> 									IntConstant[val=2](16,24)
> 	Function[isProto=0]	Void(19,0)
> 		Identifier[name:testFunctions](19,5)
> 		Scope(19,21)
> 			Scalar Declaration			Int(20,4)
> 				Identifier[name:result](20,8)
> 			Scalar Declaration			Bool(21,4)
> 				Identifier[name:evenTest](21,9)
> 			Assignment(23,4)
> 				Reference(23,4)
> 					Identifier[name:result](23,4)
> 				CallExpr(23,13)
> 					Identifier[name:factorial](23,13)
> 					Argument(23,23)
> 						Int Expression(23,23)
> 							IntConstant[val=5](23,23)
> 			Expression Statement(24,4)
> 				CallExpr(24,4)
> 					Identifier[name:writeInt](24,4)
> 					Argument(24,13)
> 						Int Expression(24,13)
> 							Reference(24,13)
> 								Identifier[name:result](24,13)
> 			Expression Statement(25,4)
> 				CallExpr(25,4)
> 					Identifier[name:newLine](25,4)
> 			Assignment(27,4)
> 				Reference(27,4)
> 					Identifier[name:evenTest](27,4)
> 				CallExpr(27,15)
> 					Identifier[name:isEven](27,15)
> 					Argument(27,22)
> 						Int Expression(27,22)
> 							IntConstant[val=4](27,22)
> 			Expression Statement(28,4)
> 				CallExpr(28,4)
> 					Identifier[name:writeBool](28,4)
> 					Argument(28,14)
> 						Int Expression(28,14)
> 							Reference(28,14)
> 								Identifier[name:evenTest](28,14)
> 			Expression Statement(29,4)
> 				CallExpr(29,4)
> 					Identifier[name:newLine](29,4)
